axon-cosmosdb/src/main/java/org/axonframework/extensions/cosmosdb/eventsourcing/tokenstore/CosmosTokenStore.java

```diff
@@ -0,0 +1,516 @@
+                        );
+                    })
+                    .block();
+        } catch (UnableToClaimTokenException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new UnableToClaimTokenException("Could not store token.", e);
+        }
+    }
+
+    @Override
+    public TrackingToken fetchToken(@Nonnull String processorName, int segment) throws UnableToClaimTokenException {
+        CosmosItemResponse<CosmosTokenItem> response = getCurrent(processorName, segment)
+                .flatMap(current -> {
+                    validateCurrent(current.getItem());
+                    return getContainer(processorName).replaceItem(
+                            current.getItem().withOwner(nodeId),
+                            current.getItem().getId(),
+                            new PartitionKey(current.getItem().getId()),
+                            new CosmosItemRequestOptions().setIfMatchETag(current.getETag())
+                    );
+                })
+                .flatMap(r -> {
+                    if (r.getStatusCode() == 200) {
+                        return getCurrent(processorName, segment);
+                    } else {
+                        return Mono.empty();
+                    }
+                })
+                .block();
+        if (response == null || response.getItem() == null) {
+            throw new UnableToClaimTokenException(NO_CURRENT_ITEM_MESSAGE);
+        }
+        return toTrackingToken(response.getItem());
+    }
+
+    @Override
+    public void extendClaim(@Nonnull String processorName, int segment) throws UnableToClaimTokenException {
+        try {
+            getCurrent(processorName, segment)
+                    .flatMap(current -> {
+                        String owner = current.getItem().getOwner();
+                        if (!nodeId.equals(owner)) {
+                            return Mono.error(new UnableToClaimTokenException(
+                                    String.format(CANT_EXTEND_MESSAGE_FORMAT, owner, nodeId)));
+                        }
+                        return getContainer(processorName).replaceItem(
+                                current.getItem().extend(),
+                                current.getItem().getId(),
+                                new PartitionKey(current.getItem().getId()),
+                                new CosmosItemRequestOptions().setIfMatchETag(current.getETag())
+                        );
+                    })
+                    .block();
+        } catch (UnableToClaimTokenException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new UnableToClaimTokenException("Could not extend token claim.", e);
+        }
+    }
+
+    @Override
+    public void releaseClaim(@Nonnull String processorName, int segment) {
+        getCurrent(processorName, segment)
+                .flatMap(current -> {
+                    validateCurrent(current.getItem());
```

Comment:
Doing the `validateCurrent` operation here means we would throw an `UnableToClaimException` if this task occurs within the `claimTimeout` window. To be honest, that seems incorrect. 

The `releaseClaim` operation is allowed to throw a warning if the release couldn't be executed, but throwing an exception is a little overboard.

Reply:
Good catch, I changed it to catch any error, and log at warn instead. Also adding a test to the integration tests.

Reply:
Awesome, thanks for that!